### ISSUE TITLE
Error in Pdfs service class when rendering a PDF

### ISSUE THREAD

**Issue by modelesque** [craftcms/commerce#2605]
### Description

We have an `Email` model set up for when an order's value changes so we can notify the customer that their order has changed. The email doesn't get sent out when the order changes its order status; it happens in the `Order::EVENT_AFTER_SAVE` event. The email renders the invoice PDF and attaches it.

Yesterday, we updated from Commerce 3.4.6 to 3.4.7 (although still on Craft 3.7.18.1) and discovered that editing a line item and updating the order throws an error in the `renderPdfForOrder()` method in `/vendor/craftcms/commerce/src/services/Pdfs.php`. It appears the method doesn't check if `$pdf` is still `null` before calling the method.

<img width="1350" alt="commerce-pdf-error" src="https://user-images.githubusercontent.com/13842870/141093902-0d81e35e-2eee-4ec7-8523-359cf1ea942b.png">

Reverting back to 3.4.6 solved the problem.

### Steps to reproduce

1. Create an `Email` that only gets rendered and sent after an order saves (and the total value of the order changes). Attach a PDF invoice in the email.
2. Edit an order's line item quantity, click "Done".
3. Update the order. The error shows upon updating.

### Additional info

- Craft CMS version: 3.7.18.1
- Craft Commerce version: 3.4.7
- PHP version: 7.4.22
- Database driver & version: MySQL 8.0.27

**Comment by pdaleramirez** [comment]
@modelesque 
I tested this in latest version and unable to reproduce it.
It seems that the issue here is that your custom plugin is calling the ` renderPdfForOrder` with a null value of $pdf variable.

**Comment by modelesque** [comment]
@pdaleramirez hmm, ok, but maybe I'm misunderstanding [the code in this method](https://github.com/craftcms/commerce/blob/4074ef2708c2f5e3e6ef98dfc870f2418d7fbe38/src/services/Pdfs.php#L448):

```
public function renderPdfForOrder(Order $order, $option = '', $templatePath = null, $variables = [], $pdf = null): string
{
    if ($pdf !== null && $pdf instanceof Pdf) {
        $templatePath = $pdf->templatePath;
    }

    if (!$templatePath) {
        $templatePath = Plugin::getInstance()->getPdfs()->getDefaultPdf()->templatePath;
    }

    // Trigger a 'beforeRenderPdf' event
    $event = new PdfEvent([
        'order' => $order,
        'option' => $option,
        'template' => $templatePath,
        'variables' => $variables,
    ]);
    $this->trigger(self::EVENT_BEFORE_RENDER_PDF, $event);

    if ($event->pdf !== null) {
        return $event->pdf;
    }

    $variables = $event->variables;
    $variables['order'] = $event->order;
    $variables['option'] = $event->option;

    // Set Craft to the site template mode
    $view = Craft::$app->getView();
    $originalLanguage = Craft::$app->language;



    ////////////////////////////////////////////////////////
    // nothing up until this point checks if $pdf is still null, which is the default for the argument
    $pdfLanguage = $pdf->getRenderLanguage($order);


    // ...etc
}
```
As my comment in the code says, the `$pdf` argument isn't required and can be `null`, but I don't see anything that checks or handles that. So maybe you can't recreate my issue exactly, but it appears possible the error I reported can happen in another scenario.

Is there something I'm not getting?

**Comment by nfourtythree** [comment]
Hi @modelesque 

What you have mentioned is true that in isolation calling this method with the `$pdf` param missing or set to `null` will result in an error. We should certainly look to refactor the method to make sure this cannot happen, ideally we would enforce the `$pdf` param.

However, as @pdaleramirez mentioned, in the normal operation of Commerce this is never an issue. There are 5 calls to the `renderPdfForOrder()` method in the Commerce code base. [One call](https://github.com/craftcms/commerce/blob/develop/src/controllers/DownloadsController.php#L73-L73) in `DownloadsController::actionPdf()`, three calls in `DownloadOrderPdfAction::actionPerform()` ([1](https://github.com/craftcms/commerce/blob/develop/src/elements/actions/DownloadOrderPdfAction.php#L121), [2](https://github.com/craftcms/commerce/blob/develop/src/elements/actions/DownloadOrderPdfAction.php#L131), [3](https://github.com/craftcms/commerce/blob/develop/src/elements/actions/DownloadOrderPdfAction.php#L148)) and [one call](https://github.com/craftcms/commerce/blob/develop/src/services/Emails.php#L706-L706) in `Emails::sendEmail()`. As you can see from this code there is always preceding checks to make sure that a PDF model is passed to the method.

This suggests that something is different in your custom code. If you would like to send your code to support@craftcms.com we could take a further look and try to decipher what might be happening.

Thanks.

**Comment by modelesque** [comment]
@nfourtythree I managed to make a fix for me, but I noticed that several updates have been published for Commerce and this method hasn't changed. I don't understand why this ticket is still closed since the main point is still relevant and you still have code that could potentially error out. 

> in the normal operation of Commerce this is never an issue.

This seems weird to me. Do you all code public methods expecting other plugin/modules to not use them? It would make more sense that your methods protect themselves and check that all the arguments are as they need to be before running the logic. Here, you're basically telling me that "we know not to pass in a null `$pdf` argument despite it being optional in the method, so all good; please adjust your code accordingly". I don't think that's the best way to handle this. It also makes me a little nervous using other Commerce methods.

**Comment by nfourtythree** [comment]
Hi @modelesque 

If you see in the first paragraph of my previous reply I did mention that what you had pointed out was true and that there is an instance where using the method in a particular way would cause an error. It was simply a case of pointing out that this is not something we had spotted due to the nature of the use in the core Commerce code.

Again, agreeing with you, I also mentioned:

> We should certainly look to refactor the method to make sure this cannot happen, ideally we would enforce the `$pdf` param.

Highlighting the fact that we should be more protective in the code so that this issue could not happen.

The issue is closed due to a resolution to the current situation having been provided. Our next step will be the refactor/update to the code which we have to make considerations around, changing a method's signature might be classed as a breaking change and this is not something we would want to do in a patch version of Commerce.

A specific issue for this refactor has been created giving you the ability to follow along with its progress there: https://github.com/craftcms/commerce/issues/2633

Thanks.
